src/celeritas/optical/CerenkovPreGenerator.hh

```diff
@@ -0,0 +1,139 @@
+
+    auto optical_dist_data = pre_cerenkov(rng);
+    if (optical_dist_data)
+    {
+        CerenkovGenerator cerenkov_gen(... , optical_dist_data, ...);
+        cerenkov_gen(rng);
+    }
+ * \endcode
+ */
+class CerenkovPreGenerator
+{
+  public:
+    // Construct with particle and material data
+    inline CELER_FUNCTION
+    CerenkovPreGenerator(ParticleTrackView const& particle_view,
+                         NativeCRef<OpticalPropertyData> const& properties,
+                         NativeCRef<CerenkovData> const& shared,
+                         OpticalMaterialId mat_id,
+                         OpticalStepCollectorData const& optical_step);
+
+    // Return a populated distribution data for the Cerenkov Generator
+    template<class Generator>
+    inline CELER_FUNCTION OpticalDistributionData operator()(Generator& rng);
+
+  private:
+    units::ElementaryCharge charge_;
+    OpticalMaterialId mat_id_;
+    OpticalStepCollectorData step_collector_data_;
+    real_type num_photons_per_len_;
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from particle, material, and optical step collection data.
+ */
+CELER_FUNCTION CerenkovPreGenerator::CerenkovPreGenerator(
+    ParticleTrackView const& particle_view,
+    NativeCRef<OpticalPropertyData> const& properties,
+    NativeCRef<CerenkovData> const& shared,
+    OpticalMaterialId mat_id,
+    OpticalStepCollectorData const& optical_step)
+    : charge_(particle_view.charge())
+    , mat_id_(mat_id)
+    , step_collector_data_(optical_step)
+{
+    CELER_EXPECT(charge_.value() != 0);
+    CELER_EXPECT(mat_id_);
+    CELER_EXPECT(step_collector_data_);
+
+    auto const& pre = step_collector_data_.points[StepPoint::pre];
+    auto const& post = step_collector_data_.points[StepPoint::post];
+    units::LightSpeed beta(real_type{0.5}
+                           * (pre.speed.value() + post.speed.value()));
+
+    CerenkovDndxCalculator dndx_calculator(
+        properties, shared, mat_id_, charge_);
+    num_photons_per_len_ = dndx_calculator(beta);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return an \c OpticalDistributionData object. If no photons are sampled, an
+ * empty object is returned and can be verified via its own operator bool.
+ *
+ * The number of photons is sampled from a Poisson distribution with a mean
+ * \f[
+   \langle n \rangle = \ell_\text{step} \frac{dN}{dx}
+ * \f]
+ * where \f$ \ell_\text{step} \f$ is the step length.
+ */
+template<class Generator>
+CELER_FUNCTION OpticalDistributionData
+CerenkovPreGenerator::operator()(Generator& rng)
+{
+    if (num_photons_per_len_ == 0)
+    {
+        return {};
+    }
+
```

Comment:
Do we have an idea of how often the number of sampled photons is zero? I wonder if we should throw out this edge case by changing the `PoissonDistribution` requirement to `lambda_ >= 0` from `lambda_ > 0`. (Then this would always result in a zero sample, so we'd throw away a random number but reduce complexity here.)

Reply:
The Poisson distribution uses a normal distribution, so we'd have to do the same there... not sure if that makes sense or not. Right now the calculated dN/dx can be zero if:
- No angle integrals (refractive index data) was provided for this optical material
- The incident particle speed is below the threshold for Cerenkov emission
- More rarely, the incident particle energy is above but very close to the threshold

Reply:
The first case is a bit odd to me as the refractive index should be a minimal requirement if the material has a optical property (necessary condition of the optical transport/physics) and therefore the angle integrals should be built if Chrenkov is one of optical processes. Okay, the extra check does not harm though.

Reply:
Maybe a naive question, but what can we assume about a material with optical properties? Should an optical material have _all_ optical properties assigned? Should it at least have refractive index (or some other minimal subset of data)? Could it just have an arbitrary subset of properties?

Reply:
I think that supporting an arbitrary subset of properties is a good strategy/design choice as a user application can be an arbitrary combination of optical processes (even with only one specific process depending on their simulation need). Concerning to the refractive index, it may be a minimal requirement as I hardly imagine that there would be a meaning optical simulation without an optical media (i.e, which by definition should have the refractive index) even though one can do whatever they want - on the other hand, as a designer/developer, you may impose your own design style and a restriction of colors.